Source/Core/VideoCommon/BPFunctions.h

```diff
@@ -7,16 +7,129 @@
 
 #pragma once
 
-#include "Common/MathUtil.h"
+#include <set>
+#include <utility>
 
-struct BPCmd;
+#include "Common/MathUtil.h"
+#include "VideoCommon/BPMemory.h"
+struct XFMemory;
 
 namespace BPFunctions
 {
+struct ScissorRange
+{
+  constexpr ScissorRange(u32 offset, u32 start, u32 end) : offset(offset), start(start), end(end) {}
+  const int offset;
+  const int start;
+  const int end;
+};
+
+struct ScissorRect
+{
+  constexpr ScissorRect(ScissorRange x_range, ScissorRange y_range)
+      :  // Rectangle ctor takes x0, y0, x1, y1.
+        rect(x_range.start, y_range.start, x_range.end, y_range.end), x_off(x_range.offset),
+        y_off(y_range.offset)
+  {
+  }
+
+  MathUtil::Rectangle<int> rect;
+  int x_off;
+  int y_off;
+
+  int GetArea() const;
+};
+
+struct ScissorResult
+{
+  ScissorResult(const BPMemory& bpmem, const XFMemory& xfmem);
+  ~ScissorResult() = default;
+  ScissorResult(const ScissorResult& other)
+      : scissor_tl{.hex = other.scissor_tl.hex}, scissor_br{.hex = other.scissor_br.hex},
+        scissor_off{.hex = other.scissor_off.hex}, viewport_left{other.viewport_left},
+        viewport_right{other.viewport_right}, viewport_top{other.viewport_top},
+        viewport_bottom{other.viewport_bottom}, m_result{other.m_result}
+  {
+  }
+  ScissorResult& operator=(const ScissorResult& other)
+  {
+    scissor_tl.hex = other.scissor_tl.hex;
+    scissor_br.hex = other.scissor_br.hex;
+    scissor_off.hex = other.scissor_off.hex;
+    viewport_left = other.viewport_left;
+    viewport_right = other.viewport_right;
+    viewport_top = other.viewport_top;
+    viewport_bottom = other.viewport_bottom;
+    m_result = other.m_result;
+    return *this;
+  }
+  ScissorResult(ScissorResult&& other)
+      : scissor_tl{.hex = other.scissor_tl.hex}, scissor_br{.hex = other.scissor_br.hex},
+        scissor_off{.hex = other.scissor_off.hex}, viewport_left{other.viewport_left},
+        viewport_right{other.viewport_right}, viewport_top{other.viewport_top},
+        viewport_bottom{other.viewport_bottom}, m_result{std::move(other.m_result)}
+  {
+  }
+  ScissorResult& operator=(ScissorResult&& other)
+  {
+    scissor_tl.hex = other.scissor_tl.hex;
+    scissor_br.hex = other.scissor_br.hex;
+    scissor_off.hex = other.scissor_off.hex;
+    viewport_left = other.viewport_left;
+    viewport_right = other.viewport_right;
+    viewport_top = other.viewport_top;
+    viewport_bottom = other.viewport_bottom;
+    m_result = std::move(other.m_result);
+    return *this;
+  }
+
+  // Input values, for use in statistics
+  ScissorPos scissor_tl;
+  ScissorPos scissor_br;
+  ScissorOffset scissor_off;
+  float viewport_left;
+  float viewport_right;
+  float viewport_top;
+  float viewport_bottom;
+
+  // Actual result
+  std::vector<ScissorRect> m_result;
+
+  ScissorRect Best() const;
+
+  constexpr bool ScissorMatches(const ScissorResult& other) const
```

Comment:
These functions marked constexpr can never actually be used in a constexpr context, since the constructors aren't constexpr